### PR TITLE
Add declaredType parameter in some DataContract Serialization methods

### LINKED ISSUE #32239: Bug in using a custom DataContractResolver
When defining a custom DataContractResolver, the following overridden function:
`public override Type ResolveName(string typeName, string typeNamespace, Type declaredType,
        DataContractResolver knownTypeResolver)`
Is always called with `declaredType = null`, when used with .net core (works fine on framework).

Attaching a solution with 2 projects - .net framework and .net core, that make a simple use of a custom DataContractResolver. The code in both is exactly the same. You can see that the problem is reproduced in the .net core project.
[DeserializationTest.zip](https://github.com/dotnet/corefx/files/2372639/DeserializationTest.zip)

### PR DESCRIPTION
**Pull request by yujayee**
Fixes #32239 
For some reason, `XmlObjectSerializerReadContext.InternalDeserialize()` methods drops `declaredType` parameter which causes the declaredType is always null when override method: `public override Type ResolveName(string typeName, string typeNamespace, Type declaredType, DataContractResolver knownTypeResolver)` Adding it in related methods.
@mconnew @huanwu @Lxiamail

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs [hunk 1]
```diff
@@ -361,7 +361,7 @@ internal void VerifyType(DataContract dataContract)
                 knownTypesAddedInCurrentScope = true;
             }
 
-            DataContract knownContract = ResolveDataContractFromKnownTypes(dataContract.StableName.Name, dataContract.StableName.Namespace, null /*memberTypeContract*/);
+            DataContract knownContract = ResolveDataContractFromKnownTypes(dataContract.StableName.Name, dataContract.StableName.Namespace, null /*memberTypeContract*/, null);
             if (knownContract == null || knownContract.UnderlyingType != dataContract.UnderlyingType)
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: You can remove this method completely. It's called by CheckIfTypeNeedsVerifcation which isn't called by anything so you can remove that method too.

**Comment by mconnew** [comment]
Can you add a unit test which fails without this fix and passes with it?

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerReadContextComplexJson.cs [hunk 2]
```diff
@@ -249,7 +249,7 @@ internal void CheckIfTypeNeedsVerifcation(DataContract declaredContract, DataCon
 
         internal void VerifyType(DataContract dataContract)
         {
-            DataContract knownContract = ResolveDataContractFromKnownTypes(dataContract.StableName.Name, dataContract.StableName.Namespace, null /*memberTypeContract*/);
+            DataContract knownContract = ResolveDataContractFromKnownTypes(dataContract.StableName.Name, dataContract.StableName.Namespace, null /*memberTypeContract*/, null);
             if (knownContract == null || knownContract.UnderlyingType != dataContract.UnderlyingType)
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerReadContextComplexJson.cs]: This method can also be removed as it's only called by CheckIfTypeNeedsVerifcation which is dead code. It doesn't exist on the .NET Framework.

**Comment by yujayee** [comment]
@dotnet-bot test Linux x64 Release Build please

**Comment by stephentoub** [comment]
@yujayee, can this be merged?

**Pull request merged**